cpp/oneapi/dal/detail/memory_impl_host.hpp

```diff
@@ -0,0 +1,52 @@
+/*******************************************************************************
+* Copyright 2020 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#pragma once
+
+#include <cstring>
+
+#include "oneapi/dal/detail/common.hpp"
+
+namespace oneapi::dal::detail {
+
+struct host_only_alloc{};
+
+template <typename T>
+inline T* malloc(const host_policy&, std::int64_t count, host_only_alloc) {
+    return new T[count];
+}
+
+template <typename T>
+inline void free(const host_policy&, T* pointer) {
+    delete[] pointer;
+}
+
+inline void memset(const host_policy&, void* dest, std::int32_t value, std::int64_t size) {
```

Comment:
A few comments from my side:
- "Implementation is not quite safe" - I agree with you, but we haven't done services to check SDL it new interfaces yet. I think this is a common problem and do not need to be solved in this PR. In my vision, we should check SDL compliance for all new sources in next PRs, when a minimal set of functionality is implemented.
- "it's always true for 32bit systems" - AFAIK, we do not support 32bit systems anymore.
- "Put TODO here to optimize it in the future" - yes, it's a good solution, I'll follow it.
- "Ideally, it should be like this" - I started just like you with templated functions with `Policy` template. But this is not necessary, so I decided to keep it as simple as possible. In my opinion, it should look like this:
```cpp
template <typename T>
inline void fill(const host_seq_policy&, T* dest, std::int64_t count, const T& value) { ... }

template <typename T>
inline void fill(const host_par_policy&, T* dest, std::int64_t count, const T& value) { ... }

// somewhere in another file

template <typename T>
inline void fill(const dpcpp_policy&, T* dest, std::int64_t count, const T& value) { ... }
```
universal reference for `value` does not make sense here in my opinion, because we don't want to move it inside `fill()`, but copy to many elements of the `dest` array.

Reply:
>  We haven't done services to check SDL. I think this is a common problem and do not need to be solved in this PR.

SDL is not one time activity or automated tool, but the guidelines that should maintained throughout whole process of development. Do no create problems if you can obviously avoid them. 

> "it's always true for 32bit systems" - AFAIK, we do not support 32bit systems anymore.

It does not matter, standard defines `int` as something between `short` and `long`:
```c++
sizeof(char) <= sizeof(short) <= sizeof(int) <= sizeof(long) <= sizeof(long long)
```

I'd better stick to standard rather to assumptions about hardware.

Reply:
Argree